app/src/lib/stores/commit-status-store.ts

```diff
@@ -0,0 +1,313 @@
+   * API requests simultaneously.
+   */
+  private readonly limit = pLimit(MaxConcurrentFetches)
+
+  public constructor(accountsStore: AccountsStore) {
+    accountsStore.getAll().then(accounts => (this.accounts = accounts))
+    accountsStore.onDidUpdate(accounts => (this.accounts = accounts))
+  }
+
+  /**
+   * Called to ensure that background refreshing is running and fetching
+   * updated commit statuses for active subscriptions. The intention is
+   * for background refreshing to be active while the application is
+   * focused.
+   *
+   * Remarks: this method will do nothing if background fetching is
+   *          already active.
+   */
+  public startBackgroundRefresh() {
+    if (this.backgroundRefreshHandle === null) {
+      this.backgroundRefreshHandle = window.setInterval(
+        () => this.queueRefresh(),
+        BackgroundRefreshInterval
+      )
+      this.queueRefresh()
+    }
+  }
+
+  /**
+   * Called to ensure that background refreshing is stopped. The intention
+   * is for background refreshing to be active while the application is
+   * focused.
+   *
+   * Remarks: this method will do nothing if background fetching is
+   *          not currently active.
+   */
+  public stopBackgroundRefresh() {
+    if (this.backgroundRefreshHandle !== null) {
+      window.clearInterval(this.backgroundRefreshHandle)
+      this.backgroundRefreshHandle = null
+    }
+  }
+
+  private queueRefresh() {
+    if (!this.refreshQueued) {
+      this.refreshQueued = true
+      setImmediate(() => {
+        this.refreshQueued = false
+        this.refreshEligibleSubscriptions()
+      })
+    }
+  }
+
+  /**
+   * Looks through all active commit status subscriptions and
+   * figure out which, if any, needs to be refreshed from the
+   * API.
+   */
+  private refreshEligibleSubscriptions() {
+    for (const key of this.subscriptions.keys()) {
+      // Is it already being worked on?
+      if (this.queue.has(key)) {
+        continue
+      }
+
+      const entry = this.cache.get(key)
+
+      if (entry && !entryIsEligibleForRefresh(entry)) {
+        continue
+      }
+
+      this.limit(() => this.refreshSubscription(key))
+        .catch(e => log.error('Failed refreshing commit status', e))
+        .then(() => this.queue.delete(key))
+
+      this.queue.add(key)
+    }
+  }
+
+  private async refreshSubscription(key: string) {
+    // Make sure it's still a valid subscription that
+    // someone might care about before fetching
+    const subscription = this.subscriptions.get(key)
+
+    if (subscription === undefined) {
+      return
+    }
+
+    const { endpoint, owner, name, ref } = subscription
+    const account = this.accounts.find(a => a.endpoint === endpoint)
+
+    if (account === undefined) {
+      return
+    }
+
+    const status = await API.fromAccount(account)
+      .fetchCombinedRefStatus(owner, name, ref)
+      .catch(err => null)
```

Comment:
I think this is the case I was referring to in this comment https://github.com/desktop/desktop/pull/7163#issuecomment-477201342 - if this errors we don't look for an existing value in the cache - we just update it to whatever we get here. Is it worth keeping the old value around for situations like the network being intermittent/unavailable?

Reply:
> Is it worth keeping the old value around for situations like the network being intermittent/unavailable?

Yes, I think this is a great shout, we definitely shouldn't be throwing the old value away (at least not initially).